serde/src/ser/mod.rs

```diff
@@ -261,123 +278,126 @@ pub trait Serializer {
     ///
-    /// By default, tuples are serialized as a sequence.
+    /// By default, tuple variants are serialized as tuples.
     #[inline]
     fn serialize_tuple_variant_elt<T>(&mut self, value: T) -> Result<(), Self::Error>
         where T: Serialize
     {
         self.serialize_tuple_struct_elt(value)
     }
 
-    /// Serializes a map.
+    /// Finishes serialization of a tuple variant.
     ///
-    /// Callees of this method need to construct a `MapVisitor`, which iterates through each item
-    /// in the map.
-    fn serialize_map<V>(&mut self, visitor: V) -> Result<(), Self::Error>
-        where V: MapVisitor;
+    /// By default, tuple variants are serialized as tuples.
+    #[inline]
+    fn serialize_tuple_variant_end(&mut self,
+                               _name: &'static str,
+                               _variant_index: usize,
+                               variant: &'static str,
+                               len: usize,
+                               state: Self::SeqState,
+                               ) -> Result<(), Self::Error> {
+        self.serialize_tuple_struct_end(variant, len, state)
+    }
 
-    /// Serializes a map element (key-value pair).
-    fn serialize_map_elt<K, V>(&mut self, key: K, value: V) -> Result<(), Self::Error>
-        where K: Serialize,
-              V: Serialize;
+    /// Serialize a map.
+    fn serialize_map(&mut self, len: Option<usize>) -> Result<Self::MapState, Self::Error>;
+
+    /// Serialize a map element
+    fn serialize_map_elt<K, V>(&mut self, key: K, value: V) -> Result<(), Self::Error> where K: Serialize, V: Serialize;
+
+    /// Finishes serializing a map
+    fn serialize_map_end(&mut self, len: Option<usize>, state: Self::MapState) -> Result<(), Self::Error>;
 
     /// Serializes a struct.
     ///
     /// By default, structs are serialized as a map with the field name as the key.
     #[inline]
-    fn serialize_struct<V>(&mut self,
-                           _name: &'static str,
-                           visitor: V) -> Result<(), Self::Error>
-        where V: MapVisitor,
+    fn serialize_struct(&mut self,
+                        _name: &'static str,
+                        len: usize,
+                        ) -> Result<Self::MapState, Self::Error>
     {
-        self.serialize_map(visitor)
+        self.serialize_map(Some(len))
     }
 
-    /// Serializes an element of a struct.
+    /// Serialize a struct field
     ///
-    /// By default, struct elements are serialized as a map element with the field name as the key.
-    #[inline]
-    fn serialize_struct_elt<V>(&mut self,
-                               key: &'static str,
-                               value: V) -> Result<(), Self::Error>
-        where V: Serialize,
-    {
+    /// By default, structs are serialized as a map with the field name as the key.
+    fn serialize_struct_elt<K, V>(&mut self, key: K, value: V) -> Result<(), Self::Error> where K: Serialize, V: Serialize {
```

Comment:
Why did you turn `key` from `&'static str` into `K`?

Separate question: do we need it to be `'static` or is `&str` good enough?